addons/im_livechat/controllers/chatbot_script.py

```diff
@@ -0,0 +1,100 @@
+# -*- coding: utf-8 -*-
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+
+from odoo import http
+from odoo.http import request
+from odoo.tools import is_html_empty, plaintext2html
+
+
+class LivechatChatbotScriptController(http.Controller):
+    @http.route('/im_livechat/chatbot_restart', type="json", auth="public", cors="*")
+    def im_livechat_chatbot_restart(self, channel_uuid, chatbot_script_id):
+
+        mail_channel = request.env['mail.channel'].sudo().search([('uuid', '=', channel_uuid)], limit=1)
+        chatbot = request.env['chatbot.script'].browse(chatbot_script_id)
+        if not mail_channel or not chatbot.exists():
+            return False
+
+        posted_message = mail_channel._chatbot_restart(chatbot)
+        return posted_message.message_format()[0] if posted_message else None
+
+    @http.route('/im_livechat/chatbot_post_welcome_messages', type="json", auth="public", cors="*")
+    def im_livechat_chatbot_post_welcome_messages(self, channel_uuid, chatbot_script_id):
+        mail_channel = request.env['mail.channel'].sudo().search([('uuid', '=', channel_uuid)], limit=1)
+        chatbot = request.env['chatbot.script'].sudo().browse(chatbot_script_id)
+        if not mail_channel or not chatbot.exists():
+            return False
+
+        return chatbot._post_welcome_steps(mail_channel).message_format()
+
+    @http.route('/im_livechat/chatbot_save_answer', type="json", auth="public", cors="*")
+    def im_livechat_chatbot_save_answer(self, channel_uuid, message_id, selected_answer_id):
+        mail_channel = request.env['mail.channel'].sudo().search(
+            [('uuid', '=', channel_uuid)], limit=1)
+        chatbot_message = request.env['chatbot.message'].sudo().search(
+            [('mail_message_id', '=', message_id)], limit=1)
+
+        if not mail_channel or not chatbot_message or chatbot_message.mail_channel_id != mail_channel:
+            return
+
+        chatbot_message.write({'user_script_answer_id': selected_answer_id})
+
+    @http.route('/im_livechat/chatbot_trigger_step', type="json", auth="public", cors="*")
+    def im_livechat_chatbot_trigger_step(self, channel_uuid, chatbot_script_id):
+
+        mail_channel = request.env['mail.channel'].sudo().search([('uuid', '=', channel_uuid)], limit=1)
+        if not mail_channel:
+            return False
+
+        next_step = False
+        if mail_channel.chatbot_current_step_id:
+            user_answer = next((
+                    message.body
+                    for message
+                    in mail_channel.message_ids.sorted(lambda message: message.id, reverse=True)))
```

Comment:
The formatting is horrible, and the inner parens unnecessary.

Also this raises an exception if there are no messages, is that expected? A simpler (and more reliable) version would be:
```suggestion
            user_answer = mail_channel.message_ids.sorted(lambda message: message.id)[-1:].body
```
though delegating the dereference of the `.body` to the next line might be clearer. You can also `reverse=True` and slice out `[:1]` instead but that seems unnecessary.